[non-strict export] support tensor attribute without other args

avikchaudhuri opened the pull request:
Summary: Without args we have a hard time detecting fake modes. This causes a fake mode mismatch error in non-strict (specifically, `aot_export_module`) when the module contains tensor attributes, because we create a fresh fake mode when we cannot detect one. The fix is to pass the same fake mode throughout.

Test Plan: added test

Differential Revision: D54516595

angelayi reviewed (approved):
It looks fine to me! But I would prefer if someone could take a second look 😅 
maybe @tugsbayasgalan?

avikchaudhuri commented:
> It looks fine to me! But I would prefer if someone could take a second look 😅 maybe @tugsbayasgalan?

Also constant lifting doesn't work with retracing (KP, see test above the one I added).

tugsbayasgalan commented on torch/export/_trace.py:
```diff
@@ -442,6 +443,7 @@ def _compiling_state_context():
             trace_joint=False,
             pre_dispatch=pre_dispatch,
             kwargs=fake_kwargs,
+            default_fake_mode=default_fake_mode,
```
Could we somehow put this function call under default_fake_mode context manager and use detect_fake_mode to find the mode in aot_autograd code?

avikchaudhuri replied:
What is this `default_fake_mode` context manager? Are you saying we should set some global and have `detect_fake_mode` pick it up?

tugsbayasgalan replied:
I think you can do:
with default_fake_mode:
    aot_export()

avikchaudhuri replied:
OK, this might work. Thanks. So finicky though. And what's `_reparametrize_module` doing? So many context managers...

tugsbayasgalan replied:
Yeahhh... I think you might have to do some if/else on the context managers to make it work..

tugsbayasgalan replied:
I think _reparametrize_module here makes sure that inputs and attributes have the same fake mode. So maybe you have to do this extra context manager when there is no attribute or no input?

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -2331,6 +2331,20 @@ def forward(self, x):
         self.assertEqual(len(ep.graph_signature.input_specs), 4)
         self.assertTrue(torch.allclose(ep.module()(*inp), transform.module()(*inp)))
 
+    @testing.expectedFailureRetraceability
```
Do you have the error message?

avikchaudhuri replied:
No, but I can tell you it fails the assert that looks for lifted tensor constants (it has none). I think at some point I printed the exported program and it had an input buffer instead.

avikchaudhuri commented:
alright @tugsbayasgalan PTAL

avikchaudhuri committed:
[non-strict export] support tensor attribute without other args (#121176)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/121176

Without args we have a hard time detecting fake modes. This causes a fake mode mismatch error in non-strict (specifically, `aot_export_module`) when the module contains tensor attributes, because we create a fresh fake mode when we cannot detect one. The fix is to pass the same fake mode throughout.

bypass-github-pytorch-ci-checks

Test Plan: added test

Reviewed By: angelayi

Differential Revision: D54516595
